### DISCUSSION TITLE
LSP extension diagnostics underlining is missing for Theia based application

### DISCUSSION THREAD

**Discussion by hasm1k** [eclipse-theia/theia#13332] [General] [answered]
I have developed a Visual Studio Code extension that utilizes the Language Server Protocol(LSP) to provide diagnostic messages for specific files. The extension works fine in VS Code, where the diagnostic messages are displayed in the Problems tab with proper underlining of the specified lines and positions on the file.

Now, I have integrated this extension into a Theia Electron application by starting it with the "--plugins=local-dir" option and providing the directory containing the VSIX files of the Code extension. While the diagnostic messages appear in the Problems tab of Theia, I am encountering an issue where the specified lines and positions are not being underlined.

The electron application dependencies are:
  "dependencies": {
    "@theia/core": "latest",
    "@theia/electron": "latest",
    "@theia/toolbar": "latest",
    "@theia/filesystem": "latest",
    "@theia/file-search": "latest",
    "@theia/navigator": "latest",
    "@theia/preferences": "latest",
    "@theia/workspace": "latest",
    "@theia/search-in-workspace": "latest",
    "@theia/plugin": "latest",
    "@theia/plugin-ext": "latest",
    "@theia/plugin-ext-vscode": "latest",
    "@theia/vsx-registry": "latest",
    "@theia/markers": "latest",
    "@theia/messages": "latest"
    }

**Comment by msujew** [comment]
Hey @hasm1k, can you provide a reproducible example? Note that I would recommend to pin the `@theia/*` dependencies to a specific version. Using `latest` might lead to mismatching version issues down the line.

I can't reproduce this issue with a LSP based vscode extension:

![image](https://github.com/eclipse-theia/theia/assets/4377073/febe09df-36ca-4642-87e5-37006e55f514)

**Reply by hasm1k** [reply]
I can provide LSP JSON which my own extension's server sends to the client to display diagnostics. Here it is:

[Trace - 4:10:28 AM] Received notification 'textDocument/publishDiagnostics'.
Params: {
"diagnostics": [
{
"message": "invalid command "idk"",
"range": {
"end": {
"character": 11,
"line": 0
},
"start": {
"character": 0,
"line": 0
}
},
"severity": 1,
"source": "x"
}
],
"uri": "file:///C%3A\example.x"
}

Am I missing any mandatory fields in the JSON for Theia to display underlining, or do I need to add new dependencies to enable underlining?

**Reply by msujew** [reply]
Theia doesn't support the language server protocol directly - all translation between LSP <-> vscode is done via the extension itself likely using the `vscode-languageclient` library. So whatever you send to the client should be interpreted by vscode and Theia in the same way.

Can you test whether your vscode extension is working as expected in our [example app](https://github.com/eclipse-theia/theia/tree/master/examples/electron)? The only other thing I can ask for is a reproducible example. As you can imagine, your issue seems to be very specific and not easily reproducible, so without an example that allows me to look where things might go wrong, there's not really a way to help you.

**Comment by hasm1k** [comment]
Hi @msujew ,
The investigations showed  that the issue is coming from the uri  that was sent for diagnostics. Client receives the following response for publishDiagnostis:
Params: {
"diagnostics": [
  ...... 
],
**"uri": "file:///C%3A\\example.x"**
}

where  we have double forward slashes in file path. My extension should be supported on both Linux and Windows platforms so the LSP servers sends preferred path. On Windows the preferred path separator is the forward slash. The second forward slash was added after converting the json-rpc to string to indicate that it is not a special character. But seems that from client side the uri still contains the double forward slashes that's why I don't see the diagnostics squiggles in editor.  Does this mean that client does not remove the special characters from "uri" field ?  (I noticed that it removes all special characters from message field  before showing it on Problems tab.)

Is this a know issue ? are there any plans to fix this issue from client side ?  Is there any workaround that I can apply ? 
Please note that  I don't see any issue on VSCode side. This is only on theia for my custom vs-code extension.

**Reply by msujew** [reply, marked as answer]
> "uri": "file:///C%3A\example.x"

Oh, I see. I probably should have noticed that in the in the previous message you sent. The issue isn't the URI escaping, but something else. Backslashes `\` are interpreted as literal backslashes per the URI spec, not path separators. Your language server should return `/` instead. This isn't a Theia issue.

**Reply by hasm1k** [reply]
Ok, I will fix it on extension side. 
@msujew I appreciate your responses, they have been incredibly helpful. Thank you.
